# Amy Campbell
# 08/2020

# INPUTS
#########
# snp-dists.tsv : snp distances from the core gene alignments 
#                 output by PRANK (via Roary) generated using Run_Roary.sh 
#                 and snp-dist,
# DFU_Staph_aureus_isolates.csv : some metadata on the isolates/ wounds themselves
#
# BigFrame.csv : CSV generated by UnmappedSequences.R which for each isolate, 
#                contains assembly quality stats(QUAST with no references),
#                closest reference genome,max and summed lengths of unmapped sequence 
#                from MiniMap2 pairwise alignment of each isolate , # plasmids circularized
#                by SPADES  
#                 
# DepthData.txt : Average depths across assemblies for each isolate, as 
#                 calculated from Bowtie alignments of trimmed raw reads 
#                 to the cleaned assemblies 
# BreathData.txt : Breadths of coverage AKA proportion of "closest" reference
#                  genome on the ML tree (by SNP distance) to each isolate that 
#                  is covered by alignment of the isolate's reads to the reference at 
#                  >=10X depth. (so #bases in reference with >10X mapping / total
#                  reference genome length)
#                 
# 

# Outputs list of 'unique' (nonclustered at within-118 SNPs distance) isolates for long read follow up 
# And a list of 'representatives' from clusters of genomes within 118 SNPs of one another. 
# As well as plots showing binary SNP distances 

# Criteria for selection of cluster representatives
###################################################
# (1) Each cluster must contain isolates from the same subject as one another 
#     (so there's at least one long read reference for every subject)
# (2) 


# Set data paths
################

# More extensive metadata for the isolates (incl. things like temperature, CFUs observed, etc.)
Isolates = "DFU_Staph_aureus_isolates.csv" 

# SNP distances (for core genome alignment excluding references/outgroup)
snpdistpath="snp-dists.tsv"

Depths= read.csv("DepthData.txt", sep="\t")

Breadths = read.csv("BreadthData.txt", sep="\t")

bigframe = read.csv("BigFrame.csv")

bigframe = bigframe %>% left_join(Depths, by="DORN")
bigframe = bigframe %>% left_join(Breadths, by="DORN")

# To figure out which subject_timepoint combos have metagenomic shotgun data matched. 
LindsayMGdata = read.csv("metamapLK.csv")
LindsayMGdata$subject_timepoint = LindsayMGdata$id
LindsayMGdata$XLindsay = LindsayMGdata$X
LindsayMGdata = LindsayMGdata[c("XLindsay", "subject_timepoint")]

# 


# Metadata
###########
IsolatesInfo = read.csv(Isolates)
IsolatesInfo$DORN = paste0("DORN", IsolatesInfo$Doern.lab.bank.)
IsolatesInfo$subject_timepoint = paste(IsolatesInfo$patient_id, IsolatesInfo$visit,sep="_")
IsolatesInfo$Genome2=paste0("DORN", IsolatesInfo$Doern.lab.bank.)
IsolatesInfo$Genome1=paste0("DORN", IsolatesInfo$Doern.lab.bank.)

# SNP Distances
###############
SNPdist = read.csv(snpdistpath, sep="\t", header=1, row.names=1)
rownames(SNPdist) = colnames(SNPdist)
SNPdist$firstgenome=row.names(SNPdist)


# Make separate columns for y axis genome vs. x axis genome 
for_merge = IsolatesInfo[c("Genome1","patient_id", "visit", "subject_timepoint")]
for_merge2 = IsolatesInfo[c("Genome2","patient_id", "visit")]
# make sure these new columns are clearly tied to the Genome 2 
colnames(for_merge2) = c("Genome2", "patient_id_2", "visit2")

# 'Melt' to build a heatmap
Melted = reshape2::melt(SNPdist, id.vars=c("firstgenome"))
colnames(Melted) = c("Genome1", "Genome2", "Distance")
Melted = Melted %>% left_join(for_merge, by="Genome1")
Melted = Melted %>% left_join(for_merge2, by="Genome2")
Melted = Melted %>% mutate(Almost_Identical = (Distance < 118) )
Melted = Melted %>% mutate(Identical = (Distance==0) )

# P1 (plotting 'almost identical' (<118 snp) core genomes)
###########################################################
p1 <- ggplot(data = Melted, aes(x=Genome2, y=Genome1, fill=Almost_Identical)) + 
  geom_tile(color = "Black") + labs(fill=">99.99% single \ncore nucleotide identity") + 
  scale_fill_manual(values=c("#0072B2", "#F0D642")) +
  ggtitle("Binary Matrix Showing Isolates with <.01% SNP Distance Between Core Genomes") + 
  theme( plot.title=element_text(hjust=.5), axis.text.x=element_text(size=7, angle=90))

plotDF = data.frame(p1$data)
LevelsGenome1 = unique((plotDF %>% arrange(subject_timepoint))$Genome1)

p1$data$Genome2 = factor(p1$data$Genome2, levels=LevelsGenome1)
p1$data$Genome1 = factor(p1$data$Genome1, levels=LevelsGenome1) # 

# Making mapping to color-code isolates from common subjects into blocks
##########################################################################
subjectTimepointDF = data.frame(LevelsGenome1)
colnames(subjectTimepointDF) = c("Genome1")
MappingSubjectTimepoint_order = subjectTimepointDF %>% left_join(IsolatesInfo[c("Genome1", "subject_timepoint", "patient_id")])
Subj_tp_order = MappingSubjectTimepoint_order$subject_timepoint
subjORder=MappingSubjectTimepoint_order$patient_id
Greys=c("#D0D0D0", "#707070")
Grey_Assignments = rep(Greys, ceiling(length(unique(MappingSubjectTimepoint_order$patient_id))/2 ))
Subjects_Unique = unique(MappingSubjectTimepoint_order$patient_id)
SubjectColor=MappingSubjectTimepoint_order$patient_id

for(i in 1:length(Subjects_Unique)){
  SubjectColor[which(SubjectColor==Subjects_Unique[i])] <- Grey_Assignments[i]
}

postscript(file="SimilarIdentityMatrix.eps", width=5000, height=5000)

p1 + geom_tile(color = "black", size=.001) +
  theme(axis.ticks.y = element_line(size=1.1, color=SubjectColor),
        axis.ticks.x = element_line(size=1.1, color=SubjectColor),
        axis.ticks.length.x = unit(1, "cm"),
        axis.ticks.length.y = unit(1, "cm"),
        plot.title=element_text(hjust=.5),
        axis.text.x=element_text(size=3, angle=90, hjust=1),
        axis.text.y=element_text(size=3, vjust=.5)) +
  scale_x_discrete(labels=subjORder) + scale_y_discrete(labels=subjORder) + coord_fixed(ratio=1)

dev.off()  

# Make plot p2 using the reordered levels from p1 but
# Fill by exact identical (0 SNP distance) similarity 
######################################################
p2 <- ggplot(p1$data, aes(x=Genome1, y=Genome2, fill=Identical)) + geom_tile(color = "black", size=.001) +
  theme(axis.ticks.y = element_line(size=1.1, color=SubjectColor),
        axis.ticks.x = element_line(size=1.1, color=SubjectColor),
        axis.ticks.length.x = unit(1, "cm"),
        axis.ticks.length.y = unit(1, "cm"),
        plot.title=element_text(hjust=.5),
        axis.text.x=element_text(size=3, angle=90, hjust=1),
        axis.text.y=element_text(size=3, vjust=.5)) +
  scale_fill_manual(values=c("#0072B2", "#F0D642")) +
  labs(fill=">Identical single \ncore nucleotide identity") + 
  scale_x_discrete(labels=subjORder) + scale_y_discrete(labels=subjORder) + coord_fixed(ratio=1) + 
  ggtitle("Binary Matrix Showing Isolates with 0 SNP Distance Between Core Genomes")

postscript(file="IdentityMatrixIdentical.eps", width=5000, height=5000)
p2
dev.off()


# Identifying clusters of almost identical core genomes (<118 snps)
###################################################################
AlmostIdenticalClusters <- list()

# Identifying clusters of 'almost almost' identical core genomes (<5 snps)
Melted = Melted %>% mutate(LessThan5_Snps = (Distance < 5))

# Filter out rows where you're comparing an isolate to itself('diagonals' in the matrix)
Melted_DifferentGenomePairs = Melted %>% filter(!(Genome1==Genome2))

# Filter to rows where you're comparing two isolates from the same subject. 
Melted_DifferentGenomePairs_SameSubj = Melted_DifferentGenomePairs %>% filter(patient_id==patient_id_2)

# <118 SNPs
AlmostIdenticalPairs = Melted_DifferentGenomePairs_SameSubj %>% filter(Almost_Identical)
length(unique(AlmostIdenticalPairs$Genome1))
Unique_AlmostIdentical=(setdiff(unique(Melted$Genome1), unique(AlmostIdenticalPairs$Genome1)))
PreviousUnique_AlmostIdentical = read.table("CompletelyUnique_99.99pctlevel.txt")
PreviousUnique_AlmostIdentical = as.character(PreviousUnique_AlmostIdentical$V1)

# Yayy the same as it was before cleaning (meaning its good that ARM has already extracted all these)
setdiff(PreviousUnique_AlmostIdentical, Unique_AlmostIdentical)
Unique_AlmostIdentical
write.table(Unique_AlmostIdentical, file="CompletelyUnique_99.99pctlevel_cleaned.txt", quote=F, sep="",row.names=F, col.names=F)

uniqueAlmostIdentical_metadata = IsolatesInfo %>% filter(DORN %in% Unique_AlmostIdentical)
length(unique(uniqueAlmostIdentical_metadata$patient_id)) # 31 subjects represented from this set.
# The other 31 must therefore be represented by cluster representatives 

# These are the isolates which differ by at least >118 SNPs from all the others 

# 0 SNPs
IdenticalPairs = Melted_DifferentGenomePairs_SameSubj %>% filter(Identical)
unpairedisolates_identical = setdiff(unique(Melted$Genome1), unique(IdenticalPairs$Genome1))
length(unpairedisolates_identical)

# <5 SNPs
AlmostAlmostIdenticalPairs = Melted_DifferentGenomePairs_SameSubj %>% filter(LessThan5_Snps)
(unique(AlmostAlmostIdenticalPairs$Genome1))
unpaired_almostalmost = setdiff(unique(Melted$Genome1), unique(AlmostAlmostIdenticalPairs$Genome1))

# Clustering for purposes of selecting representative isolates
###############################################################
AlmostIdenticalPairs_order = AlmostIdenticalPairs %>% group_by(Genome1) %>% count(Almost_Identical)
OrderingGenomesClustering = unique((AlmostIdenticalPairs_order %>% arrange(-n))$Genome1)

AlmostUnique_Genome_List=OrderingGenomesClustering
j=1
length(unique(AlmostIdenticalPairs$patient_id)) # If we just grouped by subject and did no further grouping by identity 
while(length(AlmostUnique_Genome_List) > 0){
  RepGenome = AlmostUnique_Genome_List[1]
  
  LittleList=(AlmostIdenticalPairs %>% filter(Genome1==RepGenome))$Genome2
  AlmostIdenticalClusters[[j]] <- append(LittleList, RepGenome)
  j=j+1
  AlmostUnique_Genome_List = setdiff(AlmostUnique_Genome_List, LittleList)
  AlmostUnique_Genome_List = setdiff(AlmostUnique_Genome_List, c(RepGenome))
}

# these are the clusters from which I'll choose one DORN each. 
###############################################################

# Select one from each cluster (for 44 + 42 = 86 isolates on the sequencer so far total.)
# Then, choose the remaining 9 isolates to sequence (leaving space for 1060) based on some other criteria/out of some other interest. 
bigframe = bigframe %>% left_join(IsolatesInfo[c("DORN", "subject_timepoint")], by = "DORN")
bigframe = bigframe %>% left_join(LindsayMGdata, by="subject_timepoint")
bigframe = bigframe %>% mutate(MetaGenomicData = (!is.na(XLindsay)))
bigframe$XLindsay = NULL

# The clusters are sorted by size, so the first few require more detailed attention
# Cluster 1

# Lots of timepoints in this one. It'd be interesting to have a later timepoint (e.g. 12 or 13 )
# As well as an earlier one. 
unique(bigframe1$subject_timepoint)

bigframe1 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[1]])
ggplot2::qplot(data=bigframe1, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe1$DORN) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 1 (Subject 176")
ggplot2::qplot(data=bigframe1, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe1$subject_timepoint) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 1 (Subject 176")

(bigframe1 %>% arrange(subject_timepoint))$DORN
bigframe1_subset = bigframe1 %>% filter(DORN %in% c("DORN1731", "DORN1663", "DORN1644", "DORN1885"))
selected = list()
selected[[1]] = c("DORN1885", "DORN1731")
selected[[2]] = c("DORN1849","DORN2166")
selected[[3]] = c("DORN976")



bigframe2 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[2]])
ggplot2::qplot(data=bigframe2, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe2$DORN, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
ggplot2::qplot(data=bigframe2, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe2$subject_timepoint, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
bigframe2subset = bigframe2 %>% filter(DORN %in% c("DORN2075", "DORN1849", "DORN2166"))
  
(bigframe1 %>% arrange(subject_timepoint))$DORN
selected[[2]] = c("DORN2166", "DORN1849")
bigframe3 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[3]])
bigframe3 %>% arrange(TotalLength)
ggplot2::qplot(data=bigframe3, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe3$DORN, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
ggplot2::qplot(data=bigframe3, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe3$subject_timepoint, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
selected[[3]] = c("DORN976")

bigframe4 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[4]])
bigframe4 %>% arrange(TotalLength)
ggplot2::qplot(data=bigframe4, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe4$DORN, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
ggplot2::qplot(data=bigframe4, x=BreadthOfCovg,y=Depth) + geom_text(label=bigframe4$subject_timepoint, nudge_x=-.001, nudge_y=.001) + ggtitle("Breadth of Coverage to CC5_N315 vs. Depth of \nCoverage in Cluster 2 (Subject 186)")
selected[[4]] = c("DORN2136")


# DORN 
bigframe5 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[5]])
selected[[5]] = c("DORN892", "DORN936")

bigframe6 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[6]])
bigframe6
selected[[6]] = c("DORN459")
bigframe7 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[7]])
bigframe7
selected[[7]] = c("DORN1496")


bigframe8 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[8]])

selected[[8]] = c("DORN2093")

bigframe9 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[9]])
selected[[9]] = c("DORN76", "DORN62")


bigframe10 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[10]])

selected[[10]] = c("DORN1540", "DORN1623", "DORN1546")

bigframe11 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[11]])
bigframe11
selected[[11]] = c("DORN1628", "DORN1575")
bigframe12 = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[12]])

for(i in 12:length(AlmostIdenticalClusters)){
  bigframesubset = bigframe %>% filter(DORN %in% AlmostIdenticalClusters[[i]])
  maximumunampped = max(bigframesubset$TotalLength)
  
  maxlist = which(bigframesubset$TotalLength==maximumunampped)
  if(maximumunampped == 0 | length(maxlist)> 1){
    plasmidsDF=bigframesubset[maxlist, ]
    minimumplasmids = min(plasmidsDF$NPlasmids)
    minplasmidsDF = plasmidsDF[which(plasmidsDF$NPlasmids==minimumplasmids), ]
    
    if(dim(minplasmidsDF)[1] > 1){
      winningrow  = minplasmidsDF[which(minplasmidsDF$N50==min(minplasmidsDF$N50)), ]
      selected[[i]] = c(winningrow$DORN)
      print("Based on N50")
    }else{
      winningrow  = minplasmidsDF
      selected[[i]] = c(winningrow$DORN)
      print("Based on Plasmids")
    }
      
  }else{
    print("Based on length unmapped")
    winningrow = bigframesubset[maxlist, ]
    selected[[i]] = c(winningrow$DORN)
  }
    
}

# Based on being >118 SNPs from all other isolates from the same subject
SelectedDORNs1 = Unique_AlmostIdentical

# Based on other criteria
SelectedDORNs2 = (unlist(selected))

totalSelected = c(SelectedDORNs1,SelectedDORNs2 )

write.table(SelectedDORNs2, "Cluster_Reps_ForSequencing.txt", quote=F, sep="", row.names=F, col.names=F)
write.table(totalSelected, "AllDORNs_ONP.txt", quote=F, sep="", row.names=F, col.names=F)


TotalSelectedFrame =  bigframe %>% filter(DORN %in% totalSelected)
write.csv(TotalSelectedFrame, file="SelectedONP.csv")
View(TotalSelectedFrame)
TotalSelectedFrameIsolates = IsolatesInfo %>% filter(DORN %in%totalSelected )

length(unique(TotalSelectedFrameIsolates$patient_id))
length(unique(IsolatesInfo$patient_id))

setdiff(unique(IsolatesInfo$patient_id),unique(TotalSelectedFrameIsolates$patient_id) )
View(IsolatesInfo %>% filter(patient_id %in% c(101, 307)))
# ok that we're missing 307 since its only staph aureus turned out to be staph simulans 
bigframeSubset = bigframe %>% filter(DORN %in% c("DORN5", "DORN4", "DORN6", "DORN7", "DORN8", "DORN9", "DORN14"))
# Patient 101 ?
# I guess maybe they didnt meet the new criteria? 


SelectedDORNs2

# DORN1849 too simlar to 2166 by all characteristics; 2166 has more contigs total, higher depth 
# DORN1885 Too similar to DORN1731
# DORN936 isn't a bad enough assembly to include

removal_to_make_controlspace = c("DORN1849", "DORN1623", "DORN1885","DORN62", "DORN936", "DORN1628")
SelectedDORNs2_Conservative = setdiff(SelectedDORNs2, removal_to_make_controlspace)
SelectedDORNs2_Conservative

total_selected_conservative = c(SelectedDORNs1, SelectedDORNs2_Conservative)
write.table(SelectedDORNs2_Conservative, "Cluster_Reps_ForSequencing_conservative.txt", quote=F, sep="", row.names=F, col.names=F)
write.table(total_selected_conservative, "AllDORNs87_ONP_Conservative.txt", quote=F, sep="", row.names=F, col.names=F)



unique(SelectedDORNs2)


